packages/core/src/config/parse.ts

```diff
@@ -1655,10 +1688,78 @@ const parseAndValidateChugSplashConfig = async (
     contracts: {},
   }
 
+  // Resolve all the contracts upfront so they can be used handle contract references regardless or ordering.
+  // We also cache the compiler output, constructor args, and other artifacts so we don't have to re-read them later when we complete the parsing
   const contracts = {}
+  const cachedCompilerOutput = {}
+  const cachedConstructorArgs = {}
+  const cachedArtifacts = {}
```

Comment:
> I think we can refactor the logic that retrieves address for unproxied contracts. It looks like we
> repeat this pattern of readContractArtifact -> getCreationCodeWithConstructorArgs ->
> getContractAddress in a few functions: displayDeploymentTable, parseAndValidateChugSplashConfig, and
> foundry/index.ts (under getAddress). If we refactor this into a single function, we can use a
> ternary operator for the address, which IMO is cleaner than how we do it now

If you agree with this change, I think we should remove these `cached` variables in favor of a simple ternary operator for the value of `contracts[referenceName]`. It'd make this code simpler, and I don't think it'd make it meaningfully slower given that everything's synchronous